Testscripts/Windows/Power-Hibernate.ps1

```diff
@@ -153,103 +164,122 @@ do
-			if ($state -eq 0) {
-				$kernelCompileCompleted = Run-LinuxCmd -ip $VMData.PublicIP -port $VMData.SSHPort -username $user -password $password -command "dmesg | grep -i 'hibernation exit'"
-				# This verification might be revised in future. Checking with dmesg is risky.
-				if ($kernelCompileCompleted -ne "hibernation exit") {
-					Write-LogErr "VM $($VMData.RoleName) resumed successfully but could not determine hibernation completion"
+			# Resume the VM
+			Start-AzVM -Name $vmName -ResourceGroupName $rgName -NoWait | Out-Null
+			Write-LogInfo "Waked up the VM $vmName in Resource Group $rgName and continue checking its status in every 15 seconds until 20 minutes timeout "
+
+			# Wait for VM resume for 57 min-timeout
+			$timeout = New-Timespan -Minutes 57
+			$sw = [diagnostics.stopwatch]::StartNew()
+			while ($sw.elapsed -lt $timeout) {
+				$vmCount = $AllVMData.Count
+				Wait-Time -seconds 15
+				$state = Run-LinuxCmd -ip $AllVMData[0].PublicIP -port $AllVMData[0].SSHPort -username $user -password $password -command "date > /dev/null; echo $?"
+				if ($state) {
+					$kernelCompileCompleted = Run-LinuxCmd -ip $AllVMData[0].PublicIP -port $AllVMData[0].SSHPort -username $user -password $password -command "dmesg | grep -i 'hibernation exit'"
+					# This verification might be revised in future. Checking with dmesg is risky.
+					if ($kernelCompileCompleted -ne "hibernation exit") {
+						Write-LogErr "VM $($AllVMData[0].RoleName) resumed successfully but could not determine hibernation completion"
+					} else {
+						Write-LogInfo "VM $($AllVMData[0].RoleName) resumed successfully"
+						$vmCount--
+					}
+					break
 				} else {
-					Write-LogInfo "VM $($VMData.RoleName) resumed successfully"
-					$vmCount--
+					Write-LogInfo "VM is still resuming!"
 				}
-				break
+			}
+
+			if ($vmCount -le 0) {
+				Write-LogInfo "VM resume completed"
 			} else {
-				Write-LogInfo "VM is still resuming!"
+				# Either VM hang or VM resume needs longer time.
+				throw "VM resume did not finish, the latest state was $state"
 			}
-		}
 
-		if ($vmCount -le 0){
-			Write-LogInfo "VM resume completed"
-		} else {
-			# Either VM hang or VM resume needs longer time.
-			throw "VM resume did not finish, the latest state was $state"
-		}
-		#Verify the VM status after power on event
-		$vmStatus = Get-AzVM -Name $vmName -ResourceGroupName $rgName -Status
-		if ($vmStatus.Statuses[1].DisplayStatus -eq "VM running") {
-			Write-LogInfo "$($vmStatus.Statuses[1].DisplayStatus): Verified successfully VM status is running after resuming"
-		} else {
-			Write-LogErr "$($vmStatus.Statuses[1].DisplayStatus): Could not find the VM status after resuming"
-			throw "Can not identify VM status after resuming"
-		}
+			# Verify the VM status after VM is accessible.
+			# Read VM status from the host during 10 min-timeout
```

Comment:
This segment of code repeated many times in you code, I suggest you put it into a common function.

Reply:
I see the same point, but not this time.

Reply:
Especially the kernel compilation part can be in common lib but not utilis.sh.